src/workerd/io/worker-interface.capnp

```diff
@@ -147,6 +186,277 @@ struct Trace @0x8e8d911203762d34 {
   executionModel @25 :ExecutionModel;
   # the execution model of the worker being traced. Can be stateless for a regular worker,
   # durableObject for a DO worker or workflow for the upcoming Workflows feature.
+
+  # The following structs are used only in Streaming Traces.
+
+  # The Onset struct is always sent as the first event in any trace stream. It
+  # contains general metadata about the pipeline that is being traced.
+  struct Onset {
+    accountId @0 :UInt32;
+    stableId @1 :Text;
+    scriptName @2 :Text;
+    scriptVersion @3 :ScriptVersion;
+    dispatchNamespace @4 :Text;
+    scriptId @5 :Text;
+    scriptTags @6 :List(Text);
+    entrypoint @7 :Text;
+    executionModel @8 :ExecutionModel;
+
+    # Any additional arbitrary metadata that should be associated with the onset.
+    # These are different from the tags that appear in the StreamEvent structure
+    # in that those are considered unique for each event in the stream, whereas
+    # these are considered part of the onset metadata itself, just like any of
+    # the fields above. The goal is to make Onset extensible without requiring
+    # schema changes.
+    tags @9 :List(Tag);
+  }
+
+  # The Outcome struct is always sent as the last event in any trace stream. It
+  # contains the final outcome of the trace including the CPU and wall time for
+  # the entire trace steam.
+  struct Outcome {
+    outcome @0 :EventOutcome;
+
+    # Any additional arbitrary metadata that should be associated with the outcome.
+    tags @1 :List(Tag);
+  }
+
+  # An outcome information object that describes additional detail about the outcome
+  # of an Actor/Durable object. Used primarily to identify the ActorFlushReason.
+  struct ActorFlushInfo {
+    enum Common {
+      reason @0;
+      broken @1;
+    }
+    tags @0 :List(Tag);
+  }
+
+  # A detail event indicating a subrequest that was made during a request. This
+  # can be a fetch subrequest, an RPC subrequest, a call out to a KV namespace,
+  # etc.
+  # TODO(streaming-trace): This needs to be flushed out more.
+  struct Subrequest {
+    # A monotonic sequence number that is unique within the tail. The id is
+    # used primarily to correlate with the SubrequestOutcome.
+    id @0 :UInt32;
+    info :union {
+      none @1 :Void;
+      fetch @2 :FetchEventInfo;
+      jsRpc @3 :JsRpcEventInfo;
+    }
+  }
+
+  struct SubrequestOutcome {
+    id @0 :UInt32;
+    info :union {
+      none @1 :Void;
+      fetch @2 :FetchResponseInfo;
+      custom @3 :List(Tag);
+    }
+    # A Subrequest is really a specialist kind of span, so it can have an outcome.
+    # just like a span. Unlike regular spans tho, they cannot be transactional,
+    # and are not part of the normal span stack.
+    outcome @4 :Span.SpanOutcome;
+  }
+
+  # A span event is sent only at the completion of a span, and includes markers
```

Comment:
Signaling only the completion (and not the start) of the span seems confusing. Children of the span will usually arrive before the span itself is known. Until the span is known, its _parent_ cannot be known, so these children are total orphans until all their ancestors complete. What if the system crashes before spans end?

Reply:
The start of the span *is* signaled by the first event using a previously un-seen span id. Each of these events specifies not only their span id but their parent span id so it would be possible to construct the graph of spans even without an ending span event. e.g.

```
Onset (Span 0, Parent 0)    -- Starts Span 0
Event 1 (Span 1, Parent 0)    -- Starts Span 1
Event 2 (Span 2, Parent 1)    -- Starts Span 2
Event 3 (Span 2, Parent 1)  
<crash>
```

Or, let's suppose Span 2 is created before Span 1 emits any events...

```
Onset (Span 0, Parent 0)    -- Starts Span 0
Event 1 (Span 2, Parent 1)    -- Starts Span 1 and Span 2
Event 2 (Span 2, Parent 1)  
<crash>
```